File: mcs-java/src/main/java/anc/algorithms/Genetic.java
package anc.algorithms;

import static com.google.common.collect.Lists.newArrayList;
import static java.lang.String.format;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import anc.models.GeneticConfiguration;
import anc.models.Graph;
import anc.models.MaximumCommonSubgraph;
import anc.models.Pair;

public class Genetic implements MaximumCommonSubgraphAlgorithm {
    static final int SEED = 13;
    static final Random random = new Random(SEED);

    private final int nGenomes;
    private final int nGenerations;
    private final int nCb;
    private final int nTribes;
    private final double pCrossover;
    private final double pMutation;

    public Genetic(int nGenomes, int nGenerations, int nCb, int nTribes, double pCrossover, double pMutation) {
        if (nGenomes % nTribes != 0) {
            throw new RuntimeException(format("Number of genomes %s must be devisable by the number of tribes %s!",
                    nGenerations, nTribes));
        }

        this.nGenomes = nGenomes;
        this.nGenerations = nGenerations;
        this.nCb = nCb;
        this.nTribes = nTribes;
        this.pCrossover = pCrossover;
        this.pMutation = pMutation;
    }

    public static Genetic fromConfiguration(GeneticConfiguration conf) {
        return new Genetic(conf.nGenomes, conf.nGenerations, conf.nCb, conf.nTribes, conf.pCrossover, conf.pMutation);
    }

    @Override
    public MaximumCommonSubgraph compute(Graph g1, Graph g2) {
        if (g1.n > g2.n) {
            throw new RuntimeException("Graph 1 must not be bigger than graph 2!");
        }

        final int[][] D1 = createDistanceMatrix(g1);
        final int[][] D2 = createDistanceMatrix(g2);
        final Population population = Population.generate(nGenomes, g2.n, nTribes, D1, D2);
        final List<Integer> shuffledTribeIndeces = IntStream.range(0, population.tribeSize).boxed()
                .collect(Collectors.toList());

        Genome bestGenome = population.genomes.stream().max(Comparator.comparingDouble(genome -> genome.score)).get();

        for (int generation = 0; generation < nGenerations; generation++) {
            // Shuffling
            if (generation % nCb == 0) {
                population.shuffleAll();
            } else {
                population.shuffleWithinTribe();
            }

            // Crossbreading
            for (int tribe = 0; tribe < nTribes; tribe++) {
                Collections.shuffle(shuffledTribeIndeces, random);
                for (int i = 1; i < population.tribeSize; i += 2) {
                    if (random.nextDouble() > pCrossover) {
                        continue;
                    }
                    final int tribeStart = tribe * population.tribeSize;
                    final int index1 = tribeStart + shuffledTribeIndeces.get(i - 1);
                    final int index2 = tribeStart + shuffledTribeIndeces.get(i);

                    int indexWithSmallerScore = index1;
                    int indexWithBiggerScore = index2;
                    if (population.get(index1).score > population.get(index2).score) {
                        indexWithSmallerScore = index2;
                        indexWithBiggerScore = index1;
                    }

                    final List<Genome> childeren = population.get(index1).crossover(population.get(index2)).stream()
                            .sorted(Comparator.comparingDouble(g -> g.score)).collect(Collectors.toList());

                    population.set(indexWithSmallerScore, childeren.get(childeren.size() - 2));
                    population.set(indexWithBiggerScore, childeren.get(childeren.size() - 1));

                    if (bestGenome.score < population.get(indexWithBiggerScore).score) {
                        bestGenome = population.get(indexWithBiggerScore).copy();
                    }
                }
            }

            // Mutating
            for (Genome genome : population.genomes) {
                if (random.nextDouble() <= pMutation) {
                    genome.mutate();
                    if (bestGenome.score < genome.score) {
                        bestGenome = genome.copy();
                    }
                }
            }
        }

        final List<Pair> M = reduceFunction(D1, D2, bestGenome.genes).stream()
                .sorted(Comparator.comparingInt(p -> p.v1)).collect(Collectors.toList());
        return new MaximumCommonSubgraph(M, g1, g2, bestGenome.genes);
    }

    static int[][] createDistanceMatrix(Graph g) {
        final int[][] D = new int[g.n][g.n];
        final int[] explored = new int[g.n];

        for (int i = 0; i < g.n; i++) {
            for (int j = 0; j < g.n; j++) {
                D[i][j] = g.n - 1;
            }
            D[i][i] = 0;
        }

        for (int i = 0; i < g.n; i++) {
            // BFS
            final Queue<Integer> queue = new LinkedList<>();
            explored[i] = i + 1;
            queue.add(i);
            while (!queue.isEmpty()) {
                final int v = queue.poll();
                for (int w = 0; w < g.n; w++) {
                    if (v == w || g.M[v][w] == 0 || explored[w] == i + 1) {
                        continue;
                    }
                    explored[w] = i + 1;
                    final int newDistance = Math.min(D[i][w], D[i][v] + 1);
                    D[i][w] = newDistance;
                    D[w][i] = newDistance;
                    queue.add(w);
                }
            }
        }

        return D;
    }

    static int fitnessFunction(int[][] D1, int[][] D2, List<Integer> genes) {
        final int n1 = D1.length;
        int sum = 0;
        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n1; j++) {
                sum += Math.abs(D1[i][j] - D2[genes.get(i)][genes.get(j)]);
            }
        }
        return sum;
    }

    static List<Pair> reduceFunction(int[][] D1, int[][] D2, List<Integer> genes) {
        final int n1 = D1.length;
        final int[] skipped = new int[n1];
        Arrays.fill(skipped, -1);
        List<Pair> M = new ArrayList<>();
        for (int skip = 0; skip < n1 - 1; skip++) {
            for (int i = 0; i < skip; i++) {
                skipped[i] = skip;
            }
            for (int i = skip; i < n1; i++) {
                int tmpSum = 0;
                for (int j = skip; j <= i; j++) {
                    if (skipped[j] == skip) {
                        continue;
                    }
                    for (int k = skip; k <= i; k++) {
                        if (skipped[k] == skip) {
                            continue;
                        }
                        tmpSum += Math.abs(D1[j][k] - D2[genes.get(j)][genes.get(k)]);
                    }
                }
                if (tmpSum > 0) {
                    skipped[i] = skip;
                }
            }
            final List<Pair> Mcandidate = new ArrayList<>();
            for (int i = 0; i < n1; i++) {
                if (skipped[i] < skip) {
                    Mcandidate.add(new Pair(i, genes.get(i)));
                }
            }
            if (M.size() < Mcandidate.size()) {
                M = Mcandidate;
            }
        }
        return M;
    }

    public static class Genome {
        public final List<Integer> genes;
        public int score;
        public final int[][] D1;
        public final int[][] D2;

        public Genome(List<Integer> genes, int[][] D1, int[][] D2, int score) {
            this.genes = newArrayList(genes);
            this.D1 = D1;
            this.D2 = D2;
            this.score = score;
        }

        public static Genome create(List<Integer> genes, int[][] D1, int[][] D2) {
            final int score = fitnessFunction(D1, D2, genes);
            return new Genome(genes, D1, D2, score);
        }

        public void mutate() {
            final int index1 = random.nextInt(genes.size());
            int index2 = random.nextInt(genes.size() - 1);
            if (index1 == index2) {
                index2 += 1;
            }
            Collections.swap(genes, index1, index2);

            final int oldScore = score;
            recalculateScore();
            if (score < oldScore) {
                // Undo swap
                Collections.swap(genes, index1, index2);
                score = oldScore;
            }
        }

        public List<Genome> crossover(Genome otherParent) {
            final int nGenes = genes.size();
            final List<Integer> childGenes1 = new ArrayList<>(Collections.nCopies(nGenes, -1));
            final List<Integer> childGenes2 = new ArrayList<>(Collections.nCopies(nGenes, -1));

            int nCycle = 0;
            for (int i = 0; i < nGenes; i++) {
                int j = i;
                boolean newCycle = false;
                while (childGenes1.get(j) < 0) {
                    newCycle = true;
                    if (nCycle % 2 == 0) {
                        childGenes1.set(j, genes.get(j));
                        childGenes2.set(j, otherParent.genes.get(j));
                    } else {
                        childGenes1.set(j, otherParent.genes.get(j));
                        childGenes2.set(j, genes.get(j));
                    }
                    j = genes.indexOf(otherParent.genes.get(j));
                }
                if (newCycle) {
                    nCycle += 1;
                }
            }

            return newArrayList(this, otherParent, Genome.create(childGenes1, D1, D2),
                    Genome.create(childGenes2, D1, D2));
        }

        public Genome copy() {
            return new Genome(genes, D1, D2, score);
        }

        private void recalculateScore() {
            this.score = fitnessFunction(D1, D2, genes);
        }
    }

    public static class Population {
        public final List<Genome> genomes;
        public final int nTribes;
        public final int tribeSize;

        public Population(List<Genome> genomes, int nTribes) {
            this.genomes = newArrayList(genomes);
            this.nTribes = nTribes;
            this.tribeSize = genomes.size() / nTribes;

            if (tribeSize % 2 == 1) {
                throw new RuntimeException(format("Tribe size %s must divisible by 2!", tribeSize));
            }
        }

        public static Population generate(int nGenomes, int genomeSize, int nTribes, int[][] D1, int[][] D2) {
            final List<Genome> genomes = new ArrayList<>(nGenomes);

            for (int i = 0; i < nGenomes; i++) {
                final List<Integer> genes = new ArrayList<>(genomeSize);
                for (int j = 0; j < genomeSize; j++) {
                    genes.add(j);
                }
                Collections.shuffle(genes, random);
                genomes.add(Genome.create(genes, D1, D2));
            }
            return new Population(genomes, nTribes);
        }

        public Genome get(int index) {
            return genomes.get(index);
        }

        public void set(int index, Genome genome) {
            genomes.set(index, genome);
        }

        public void shuffleAll() {
            Collections.shuffle(genomes, random);
        }

        public void shuffleWithinTribe() {
            for (int j = 0; j < nTribes; j++) {
                Collections.shuffle(genomes.subList(j * tribeSize, (j + 1) * tribeSize), random);
            }
        }
    }
}

File: Mcs-cs/Mcs/Algorithms/Genetic.cs
using Mcs.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Mcs.Algorithms
{
    public class Genetic : MaximumCommonSubgraphAlgorithm
    {
        private static readonly int SEED = 13;
        private static readonly Random RANDOM = new Random(SEED);

        private readonly int NGenomes;
        private readonly int NGenerations;
        private readonly int NCb;
        private readonly int NTribes;
        private readonly double PCrossover;
        private readonly double PMutation;

        public Genetic(int nGenomes, int nGenerations, int nCb, int nTribes, double pCrossover, double pMutation)
        {
            if (nGenomes % nTribes != 0)
            {
                throw new SystemException("Number of genomes must be devisable by the number of tribes!");
            }

            NGenomes = nGenomes;
            NGenerations = nGenerations;
            NCb = nCb;
            NTribes = nTribes;
            PCrossover = pCrossover;
            PMutation = pMutation;
        }

        public static Genetic FromConfiguration(GeneticConfiguration conf)
        {
            return new Genetic(conf.NGenomes, conf.NGenerations, conf.NCb, conf.NTribes, conf.PCrossover, conf.PMutation);
        }

        public MaximumCommonSubgraph Compute(Graph g1, Graph g2)
        {
            if (g1.N > g2.N)
            {
                throw new SystemException("Graph 1 must not be bigger than graph 2!");
            }

            List<List<int>> D1 = CreateDistanceMatrix(g1);
            List<List<int>> D2 = CreateDistanceMatrix(g2);
            Population population = Population.Generate(NGenomes, g2.N, NTribes, D1, D2);
            Genome bestGenome = population.Genomes.OrderBy(g => g.Score).Last();

            for (int generation = 0; generation < NGenerations; generation++)
            {
                // Shuffling
                if (generation % NCb == 0)
                {
                    population.ShuffleAll();
                }
                else
                {
                    population.ShuffleWithinTribe();
                }

                // Crossbreading
                for (int tribe = 0; tribe < NTribes; tribe++)
                {
                    for (int i = 1; i < population.TribeSize; i += 2)
                    {
                        if (RANDOM.NextDouble() > PCrossover)
                        {
                            continue;
                        }
                        int tribeStart = tribe * population.TribeSize;
                        int index1 = tribeStart + i - 1;
                        int index2 = tribeStart + i;

                        int indexWithSmallerScore = index1;
                        int indexWithBiggerScore = index2;
                        if (population[index1].Score > population[index2].Score)
                        {
                            indexWithSmallerScore = index2;
                            indexWithBiggerScore = index1;
                        }

                        List<Genome> childeren = population[index1]
                            .Crossover(population[index2])
                            .OrderBy(g => g.Score)
                            .ToList();

                        population[indexWithSmallerScore] = childeren[childeren.Count - 2];
                        population[indexWithBiggerScore] = childeren[childeren.Count - 1];

                        if (bestGenome.Score < population[indexWithBiggerScore].Score)
                        {
                            bestGenome = population[indexWithBiggerScore].Copy();
                        }
                    }
                }

                // Mutating
                foreach (Genome genome in population.Genomes)
                {
                    if (RANDOM.NextDouble() <= PMutation)
                    {
                        genome.Mutate();
                        if (bestGenome.Score < genome.Score)
                        {
                            bestGenome = genome.Copy();
                        }
                    }
                }
            }

            List<Pair> M = ReductionFunction(D1, D2, bestGenome.Genes);
            return new MaximumCommonSubgraph(g1, g2, M, bestGenome.Genes);
        }

        static List<List<int>> CreateDistanceMatrix(Graph g)
        {
            List<List<int>> D = new List<List<int>>(g.N);
            List<int> explored = new List<int>(g.N);
            for (int i = 0; i < g.N; i++)
            {
                explored.Add(0);
                D.Add(new List<int>(g.N));
                for (int j = 0; j < g.N; j++)
                {
                    D[i].Add(g.N - 1);
                }
                D[i][i] = 0;
            }

            for (int i = 0; i < g.N; i++)
            {
                // BFS
                Queue<int> queue = new Queue<int>();
                explored[i] = i + 1;
                queue.Enqueue(i);
                while (queue.Any())
                {
                    int v = queue.Dequeue();
                    for (int w = 0; w < g.N; w++)
                    {
                        if (v == w || g.AdjacencyMatrix[v][w] == 0 || explored[w] == i + 1)
                        {
                            continue;
                        }
                        explored[w] = i + 1;
                        int newDistance = Math.Min(D[i][w], D[i][v] + 1);
                        D[i][w] = newDistance;
                        D[w][i] = newDistance;
                        queue.Enqueue(w);
                    }
                }
            }

            return D;
        }

        static int FitnessFunction(List<List<int>> d1, List<List<int>> d2, List<int> genes)
        {
            int n1 = d1.Count;
            int sum = 0;
            for (int i = 0; i < n1; i++)
            {
                for (int j = 0; j < n1; j++)
                {
                    sum += Math.Abs(d1[i][j] - d2[genes[i]][genes[j]]);
                }
            }
            return sum;
        }

        static List<Pair> ReductionFunction(List<List<int>> d1, List<List<int>> d2, List<int> genes)
        {
            int n1 = d1.Count;
            List<int> skipped = NCopies(n1, -1);
            List<Pair> M = new List<Pair>();
            for (int skip = 0; skip < n1 - 1; skip++)
            {
                for (int i = 0; i < skip; i++)
                {
                    skipped[i] = skip;
                }
                for (int i = skip; i < n1; i++)
                {
                    int tmpSum = 0;
                    for (int j = skip; j <= i; j++)
                    {
                        if (skipped[j] == skip)
                        {
                            continue;
                        }
                        for (int k = skip; k <= i; k++)
                        {
                            if (skipped[k] == skip)
                            {
                                continue;
                            }
                            tmpSum += Math.Abs(d1[j][k] - d2[genes[j]][genes[k]]);
                        }
                    }
                    if (tmpSum > 0)
                    {
                        skipped[i] = skip;
                    }
                }
                List<Pair> Mcandidate = new List<Pair>();
                for (int i = 0; i < n1; i++)
                {
                    if (skipped[i] < skip)
                    {
                        Mcandidate.Add(new Pair(i, genes[i]));
                    }
                }
                if (M.Count < Mcandidate.Count)
                {
                    M = Mcandidate;
                }
            }
            return M;
        }

        class Genome
        {
            public readonly List<int> Genes;
            public int Score;
            public readonly List<List<int>> D1;
            public readonly List<List<int>> D2;

            public Genome(List<int> genes, List<List<int>> d1, List<List<int>> d2, int score)
            {
                Genes = new List<int>(genes);
                D1 = d1;
                D2 = d2;
                Score = score;
            }

            public static Genome Create(List<int> genes, List<List<int>> d1, List<List<int>> d2)
            {
                int score = FitnessFunction(d1, d2, genes);
                return new Genome(genes, d1, d2, score);
            }

            public void Mutate()
            {
                int index1 = RANDOM.Next(Genes.Count);
                int index2 = RANDOM.Next(Genes.Count - 1);
                if (index1 == index2)
                {
                    index2 += 1;
                }
                int oldScore = Score;

                Swap(Genes, index1, index2);
                RecalculateScore();
                if (Score < oldScore)
                {
                    // Undo swap
                    Swap(Genes, index1, index2);
                    Score = oldScore;
                }
            }

            /// <summary>
            /// Crossovers with otherParent
            /// </summary>
            /// <param name="otherParent"></param>
            /// <returns>List of four genomes: two parents and two children</returns>
            public List<Genome> Crossover(Genome otherParent)
            {
                int nGenes = Genes.Count;
                List<int> childGenes1 = NCopies(nGenes, -1);
                List<int> childGenes2 = NCopies(nGenes, -1);

                int nCycle = 0;
                for (int i = 0; i < nGenes; i++)
                {
                    int j = i;
                    bool newCycle = false;
                    while (childGenes1[j] < 0)
                    {
                        newCycle = true;
                        if (nCycle % 2 == 0)
                        {
                            childGenes1[j] = Genes[j];
                            childGenes2[j] = otherParent.Genes[j];
                        }
                        else
                        {
                            childGenes1[j] = otherParent.Genes[j];
                            childGenes2[j] = Genes[j];
                        }
                        j = Genes.IndexOf(otherParent.Genes[j]);
                    }
                    if (newCycle)
                    {
                        nCycle += 1;
                    }
                }

                return new List<Genome>() {
                    this,
                    otherParent,
                    Genome.Create(childGenes1, D1, D2),
                    Genome.Create(childGenes2, D1, D2) };
            }

            public Genome Copy()
            {
                return new Genome(Genes, D1, D2, Score);
            }

            private void RecalculateScore()
            {
                Score = FitnessFunction(D1, D2, Genes);
            }
        }

        class Population
        {
            public readonly List<Genome> Genomes;
            public readonly int NTribes;
            public readonly int TribeSize;

            public Population(List<Genome> genomes, int nTribes)
            {
                Genomes = new List<Genome>(genomes);
                NTribes = nTribes;
                TribeSize = genomes.Count / nTribes;

            }

            public static Population Generate(int nGenomes, int genomeSize, int nTribes, List<List<int>> d1, List<List<int>> d2)
            {
                List<Genome> genomes = new List<Genome>(nGenomes);

                for (int i = 0; i < nGenomes; i++)
                {
                    List<int> genes = new List<int>(genomeSize);
                    for (int j = 0; j < genomeSize; j++)
                    {
                        genes.Add(j);
                    }
                    Shuffle(genes);
                    genomes.Add(Genome.Create(genes, d1, d2));
                }
                return new Population(genomes, nTribes);
            }

            public Genome this[int index]
            {
                get => Genomes[index];
                set => Genomes[index] = value;
            }

            public void ShuffleAll()
            {
                Shuffle(Genomes);
            }

            public void ShuffleWithinTribe()
            {
                for (int i = 0; i < NTribes; i++)
                {
                    Shuffle(Genomes, i * TribeSize, (i + 1) * TribeSize);
                }
            }
        }

        private static void Shuffle<T>(IList<T> list, int start, int end)
        {
            int n = end - start;
            while (n > 1)
            {
                n--;
                int k = RANDOM.Next(n + 1);
                T value = list[start + k];
                list[start + k] = list[start + n];
                list[start + n] = value;
            }
        }

        private static void Shuffle<T>(IList<T> list)
        {
            Shuffle(list, 0, list.Count);
        }

        private static void Swap<T>(IList<T> list, int indexA, int indexB)
        {
            T tmp = list[indexA];
            list[indexA] = list[indexB];
            list[indexB] = tmp;
        }

        private static List<T> NCopies<T>(int n, T item)
        {
            List<T> list = new List<T>(n);
            for (int i = 0; i < n; i++)
            {
                list.Add(item);
            }
            return list;
        }
    }
}
